cpp/src/arrow/compute/kernels/scalar_compare.cc

```diff
@@ -156,39 +210,50 @@ struct Maximum {
   }
 };
 
+// Check if timestamp timezones are comparable (either all are empty or none is).
+Status CheckCompareTimestamps(const ExecBatch& batch) {
+  if (batch.num_values() > 0) {
+    const auto& ts0 = checked_cast<const TimestampType&>(*batch[0].type());
+    bool invalid_state = ts0.timezone().empty();
+    for (int i = 1; i < batch.num_values(); ++i) {
+      const auto& ts = checked_cast<const TimestampType&>(*batch[i].type());
+      invalid_state ^= ts.timezone().empty();
+    }
+    if (invalid_state) {
+      return Status::Invalid(
+          "Cannot compare timestamp with timezone to timestamp without timezone");
+    }
```

Comment:
Oops! I left a bug here. You need to move the check`if (invalid_state) {...}` inside the `for-loop`.
As it is right now, it will only work for binary cases not ternary or higher because `1^0^1` --> `0`.

Reply:
Fixed.